Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm

```diff
@@ -887,30 +882,45 @@ - (NSArray *)_additionalAccessibilityAttributeNames:(const RefPtr<AXCoreObject>&
     if (backingObject->supportsKeyShortcuts())
         [additional addObject:NSAccessibilityKeyShortcutsAttribute];
 
-    AccessibilitySortDirection sortDirection = backingObject->sortDirection();
-    if (sortDirection != AccessibilitySortDirection::None && sortDirection != AccessibilitySortDirection::Invalid)
-        [additional addObject:NSAccessibilitySortDirectionAttribute];
+    if (backingObject->supportsLiveRegion()) {
+        [additional addObject:NSAccessibilityARIALiveAttribute];
+        [additional addObject:NSAccessibilityARIARelevantAttribute];
+    }
 
-    // If an object is a child of a live region, then add these
-    if (backingObject->isInsideLiveRegion())
-        [additional addObject:NSAccessibilityARIAAtomicAttribute];
+    bool addedHasPopup = false;
+    if (backingObject->hasPopup()) {
+        addedHasPopup = true;
+        [additional addObject:NSAccessibilityHasPopupAttribute];
+    }
+    bool addedAriaAtomic = false;
+    bool addedSortDirection = false;
+    for (RefPtr ancestor = backingObject; ancestor; ancestor = ancestor->parentObject()) {
+        if (!addedHasPopup && ancestor->isLink() && ancestor->hasPopup()) {
```

Comment:
To avoid needing to keep behavior in sync, should we consider changing `selfOrAncestorLinkHasPopup` to return an `AXCoreObject` pointer that we can bool/null check?

Reply:
Hmm, it would be great if we could avoid needing to keep the behavior in sync manually (though I think it's a fine cost to pay considering how hot this function is, if we can't figure something out).

You're suggesting something like `selfOrAncestorLinkWithPopup`. How would we use that here? If we just called it outright, we would do an ancestry traversal as part of that function, and then do another to try to find aria-atomic and aria-sort ancestors, and duplicated ancestry traversals is what this PR is trying to avoid since this function can be so frequently called by ATs. But maybe you have a different idea for using this that I'm missing!

Reply:
That's a good point—the diff wasn't clear to me that those were all contained within this traversal.

I don't think there is another option then to avoid manually syncing these. What you have now is fine